### DISCUSSION TITLE
Block multiples accesses from the same login

### DISCUSSION THREAD

**Discussion by molimat** [supabase/supabase#955] [Questions]
As the title says, currently I could do multiple logins with my e-mail. How do I prevent that with supabase?

**Comment by silentworks** [comment]
What do you mean by multiple logins with your email? do you mean you can register multiple accounts with your email or do you mean you can log into the same account on multiple computers? If the latter, that is not a security issue as such since this is possible in almost all publicly accessible systems out there except for companies who have enterprise security requirements that don't normally make sense.

Also the latter is not preventable through supabase itself at the moment.

**Reply by molimat** [reply]
its the latter, it's to prevent account sharing.

**Reply by silentworks** [reply]
This isn't a simple issue to fix, now imagine I have logged in from my desktop and had to change room within the hour so I decide to log into my laptop (I'm still logged into my desktop btw), the system would lock me out due to me still being logged into my desktop. This doesn't seem like a robust system to me, remember supabase is using JWT when it logs in and I think the token is valid for 1 hour, currently I don't think there is a way to invalidate those tokens. You could build your own system on top of supabase where you keep an audit trail of logins and log a user back out if you can see that they were already logged in, in the past hour. This would mean you would have to create a manual check from your app to check if the user is already logged in and then log them out using the supabase api.

**Comment by itechnica** [comment]
I have seen this done quite neatly, although I did not code the product that does it, so I can only give you an outline of how it works...

User signs-in on DeviceA and uses the application. Application stores a token on the server, saying "DeviceA is active".

Same user signs-in on DeviceB, and the application checks the token on server, and advises user; "You are signed-in on another device, do you want to deactivate the other device and use this one"? If user says, "Yes", the token for DeviceA is marked as deactivated on the server, and user now uses DeviceB, which updates the token to say, "DeviceB is active".

If user accesses DeviceA again, it notifies them that their device is currently deactivated because the account is active on another device; do they want to deactivate the other device and use this one? The user can say yes, thus reactivating DeviceA without having to sign-in again, but now DeviceB is deactivated, and so-on.

This has been working successfully with thousands of users for nearly a year, and we haven't had a single complaint. The intent is to allow users who are 'out-in-the-field' to switch between multiple devices such as their laptop, tablet and phone, without having to keep signing in and out. For example, they use their laptop when sitting in their car or cafe, because it's convenient to use a full-size keyboard, then they use their tablet or phone to add information into the application whilst on the client's premises... just adding a couple of fields is easy enough with an on-screen keyboard.

Naturally, this only prevents simultaneous sharing of the account, because every time another person tried to use the same account, it would deactivate other devices that are using the same credentials. If two people were in opposing time zones, they could each use the same account during the day!  However, it makes casual sharing in the same time zone very inconvenient.

**Reply by NixBiks** [reply]
Can you elaborate on the technical side how you did this?

**Reply by benjaminlewandowski** [reply]
i would also like to read a more in depth explanation of the technical side of this approach

**Comment by galacticgibbon** [comment]
A cheap way to reduce account sharing would just be to restrict auth to non password based auth - google oauth, magic links etc. so that the email account owner has to help them sign on. Not exactly a huge deterrent but would likely reduce it.
